codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java

```diff
@@ -0,0 +1,203 @@
+ * should improve the goodput on each written stream.
+ */
+public final class UniformStreamByteDistributor implements StreamByteDistributor {
+    static final int DEFAULT_MIN_ALLOCATION_CHUNK = 1024;
+
+    private final Http2Connection.PropertyKey stateKey;
+    private final Deque<State> queue = new ArrayDeque<State>();
+    private final EnqueueStrategy defaultEnqueueStrategy = new EnqueueStrategy();
+
+    /**
+     * The first time a state is enqueued, it will be added to the head of the queue. After
+     * that it will be added to the tail.
+     */
+    private final EnqueueStrategy firstEnqueueStrategy = new EnqueueStrategy() {
+        @Override
+        void enqueue(State state) {
+            queue.addFirst(state);
+
+            // Switch to the default strategy so that it will be added to the tail from
+            // now on.
+            state.enqueueStrategy = defaultEnqueueStrategy;
+        }
+    };
+
+    /**
+     * The minimum number of bytes that we will attempt to allocate to a stream. This is to
+     * help improve goodput on a per-stream basis.
+     */
+    private int minAllocationChunk = DEFAULT_MIN_ALLOCATION_CHUNK;
+    private long totalStreamableBytes;
+
+    public UniformStreamByteDistributor(Http2Connection connection) {
+        checkNotNull(connection, "connection");
+
+        // Add a state for the connection.
+        stateKey = connection.newKey();
+        Http2Stream connectionStream = connection.connectionStream();
+        connectionStream.setProperty(stateKey, new State(connectionStream));
+
+        // Register for notification of new streams.
+        connection.addListener(new Http2ConnectionAdapter() {
+            @Override
+            public void onStreamAdded(Http2Stream stream) {
+                stream.setProperty(stateKey, new State(stream));
+            }
+
+            @Override
+            public void onStreamClosed(Http2Stream stream) {
+                state(stream).close();
+            }
+        });
+    }
+
+    /**
+     * Sets the minimum allocation chunk that will be allocated to each stream. Defaults to 1KiB.
+     *
+     * @param minAllocationChunk the minimum number of bytes that will be allocated to each stream.
+     * Must be > 0.
+     */
+    public void minAllocationChunk(int minAllocationChunk) {
+        if (minAllocationChunk <= 0) {
+            throw new IllegalArgumentException("minAllocationChunk must be > 0");
+        }
+        this.minAllocationChunk = minAllocationChunk;
+    }
+
+    @Override
+    public void updateStreamableBytes(StreamState streamState) {
+        State state = state(streamState.stream());
+        state.updateStreamableBytes(streamState.streamableBytes(), streamState.hasFrame());
+    }
+
+    @Override
+    public boolean distribute(int maxBytes, Writer writer) throws Http2Exception {
+        checkNotNull(writer, "writer");
+
+        final int size = queue.size();
+        if (size == 0 || maxBytes <= 0) {
```

Comment:
It seems @Scottmitch's case is still an issue, because this method will return before getting to the for loop.

It seems "stream has empty frame to send" is continually a problem. Why don't we just send immediately instead of enqueuing in that case?

Draining and refilling the queue to find empty streams seems like an overkill. If there turns out to be some reason we can't just send such things immediately, then I propose two alternatives:

 1. Keep a separate queue of streams that only have empty frames to send, and always drain that queue in `distribute()`.
 2. Just wait until you get to such streams in the queue:

    ```java
    while (!queue.isEmpty()) {
      State state = queue.remove();
      if (!(maxBytes > 0 || state.streamableBytes == 0)) {
        break;
      }
      ...
    }
    ```